include/mesh/mesh_base.h

```diff
@@ -263,6 +263,29 @@ class MeshBase : public ParallelObject
    */
   void set_elem_dimensions(const std::set<unsigned char> & elem_dims);
 
+  /**
+   * Tabulate a user-defined "code" for elements which belong to the element sets
+   * specified in \p id_set. Also sets up the inverse mapping, so that if one knows
+   * all the element sets an Elem belongs to, one can look up the corresponding code.
+   */
+  void add_elemset_code(dof_id_type code, const std::set<elemset_id_type> & id_set);
```

Comment:
Bear with me, but: are we sure we want to use `set` to store a set?  A sorted `vector` is typically faster unless the container size gets quite large or the number of insertions quite high, and in those cases `unordered_set` can be faster.

I could see this becoming a seriously useful fundamental feature, and if we're not certain what the best API is now, we might want to wrap it in some `ElemSets` class so that we can change the data structure later.

Reply:
My issue with using a vector is that you are always having to remember to keep it sorted and unique. I don't expect the construction of sets for the purpose of calling `add_elemset_code()` to be a performance bottleneck, so I'm leaning toward the convenience and familiarity of plain old `std::set` currently.

Reply:
It's not just the set construction (that actually works better with `set` than with sorted vectors), it's the lookups - a binary search in a sorted vector can go much faster just from fewer cache misses.

But I can't argue with convenience.  The best alternative I could propose would be to bring Boost's `flat_set` into our `boostcopy` namespace (or into our contrib boost, and then just have our shim fall back on `std::set` for people who disable boost).

Reply:
I was not previously familiar with `boost::flat_set`, but it sounds like the best of both worlds. According to some [Benchmarks](https://denisyaroshevskiy.github.io/benchmarking_sets/), for range construction, copying, and erasing entries, boost::flat_set kills std::set and oftentimes std::unordered_set over the entire range of sizes tested. So I think it would be good to have `boost::flat_set` available regardless of whether we end up using it for this specifically. In this PR I will typedef the type of `id_set` (for now to `std::set`) to make it easier to change later.

Reply:
Sounds like a good compromise - future changes would be technically API-breaking but we'll have given user code a chance at staying forward compatible..  Should we tag this (the add method, I guess) with `libmesh_experimental()` for a while too?

Reply:
> Should we tag this (the add method, I guess) with libmesh_experimental() for a while too?

Yeah that's fine with me.